keps/349-all-or-nothing/README.md

```diff
@@ -0,0 +1,485 @@
+
+**Component level cons**:
+- throughput - would degrade substantially by sequencing of all workloads, even those which don't require the semantics
+
+### Risks and Mitigations
+
+If the workload fails to schedule its pods it could block admission of other
+workloads indefinitely.
+
+To mitigate this issue we introduce a timeout on reaching the `PodsReady`
+condition by a workload since its job start (see:
+[Timeout on reaching the PodsReady condition](#timeout-on-reaching-the-podsready-condition)).
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### PodsReady workload condition
+
+First, we introduce a new workload condition, called `PodsReady`, to indicate
+that the first batch of a workload's pods is scheduled, more precisely:
+`job.status.ready + job.status.succeeded >= job.spec.parallelism`.
+
+The `PodsReady` condition is addded to the workload by the Kueue's Job
+reconciler in reaction to a status update of the corresponding Job. Note that,
+verifying if the condition should be added does not require an extra API call as
+the Kueue's Job reconciller already fetches the latest Job object at the
+beginning of the `Reconcile` function. Also note, there is no new condition
+introduced to the Job API.
+
+Note that, we don't take failed pods directly into the computation for verifying
+the `PodsReady` condition. However, a buggy workload is eliminated as the
+correspoding job fails due to exceeding the `.spec.backoffLimit` limit.
+
+### Workload API
+
+We extend the Workload API with the new `startupRequirements` field to
+enable the new behavior.
+
+```golang
+// StartupRequirements describes workload startup requirements
+type StartupRequirements string
+
+const (
+  // QuotaAndPodsReady: means that workload needs to get the ClusterQueue
+  // quota to be admitted and to have all its pods running at the same time
+  StartupRequirementsQuotaAndPodsReady StartupRequirements = "QuotaAndPodsReady"
+  // QuotaOnly means that the only requirement for a workload to start is to
+  // get resources from the ClusterQueue quota
+  StartupRequirementsQuotaOnly StartupRequirements = "QuotaOnly"
```

Comment:
With current naming, I find the API suspicious conceptually. I think we would rather want to assume that all jobs want atomic start. Or at least we already assume that with quota. Consider that QuotaOnly is artificial -- if the job is okay with partial start, why wouldn't quota allow that? One day we might want a true option for partial start (e.g. # min pods to start), but quota should respect it too.

So I think it's not really a Job option, but a temporary tradeoff in Kueue -- we'd like to help jobs with atomic expectations (all jobs with current Kueue assumptions), but the only way for now is to slow down to sequential admission. Since it's slower, we offer it as an option, e.g. ensure-pods-start=true. Once we have the efficient parallel solution for atomic provisioning (Resource Reservation), we would likely want to make it default. In any case, I don't see much space conceptually for QuotaOnly in between partial and atomic jobs.

Reply:
I am on the fence on this as well, I made a similar comment on the initial proposal. The relevant workload/job API is whether or not the job wants all-or-nothing scheduling. 

A global option is going to be too strict. Consider a cluster with two types of resources: reservations and spot. Jobs admitted on reservations don't need to ensure that pods are ready, while ones admitted on spot will want to due to the potential of stock outs. This issue however will not be solved with the current workload-level API (because CQ could have multiple flavors that a workload can fallback to), it would be solved if the API is part of the resourceFlavor one.

That being said, I think it is safer to start with a global option, it will not be too hard to extend it to either a workload-level API or a resourceFlavor one.

Reply:
I have renamed the option and the field documentation for `startupRequirements`, to take into account the point about quota. Current proposal: `AllPodsReady, `AtLeastOnePodReady`. 

I think generally it is a workload property - some ML workloads require all pods running at the same time, while other may require a single pod - it would just be slow.

Reply:
I haven't caught up with all the discussions yet, but I don't like Job.StartUpRequirements=AtLeastOnePodReady if "at least one" isn't respected by quota (job waiting on quota although there is quota for at least one pod).

I think conceptually there are two behaviors for Jobs:
1. Partial, min!=max: Can start partially.
2. Atomic, min=max, wants all pods.

Between these two I agree that it would be Job property. But we are not trying to address (1) in this KEP. We are differentiating variants of (2) -- whether we guarantee it only on quota level (current behavior), or also after quota (new behavior). So I would vote for a more direct name, e.g. ensure-pods-start=true, and probably more global option (global/clusterqueue/localqueue/flavor).

Between global/clusterqueue/localqueue/flavor/job I'm not sure, but I would drive this by user requirements on granularity -- do we expect users to configure it differently for different clusterqueues/localqueues/flavors/jobs ? Current user stories suggest no, but per Abdullah's point maybe we want to expand stories for reservations & spot. Also agree that if we are unsure, starting with more global is simpler and safer.

Reply:
I see the issue. I've updated the proposal to opt-in for the mechanism at the Kueue configuration level for now.